modules/core/keeper/msg_server.go

```diff
@@ -410,12 +410,11 @@ func (k Keeper) RecvPacket(goCtx context.Context, msg *channeltypes.MsgRecvPacke
 	// Cache context so that we may discard state changes from callback if the acknowledgement is unsuccessful.
 	cacheCtx, writeFn = ctx.CacheContext()
 	ack := cbs.OnRecvPacket(cacheCtx, msg.Packet, relayer)
-	// NOTE: The context returned by CacheContext() refers to a new EventManager, so it needs to explicitly set events to the original context.
-	// Events from callback are emitted regardless of acknowledgement success
-	ctx.EventManager().EmitEvents(cacheCtx.EventManager().Events())
 	if ack == nil || ack.Success() {
 		// write application state changes for asynchronous and successful acknowledgements
 		writeFn()
+		// NOTE: The context returned by CacheContext() refers to a new EventManager, so it needs to explicitly set events to the original context.
+		ctx.EventManager().EmitEvents(cacheCtx.EventManager().Events())
```

Comment:
Why does this event moved?

Reply:
https://github.com/crescent-network/ibc-go/pull/3
https://github.com/sei-protocol/sei-ibc-go/pull/22

These are PRs that apply https://github.com/cosmos/ibc-go/pull/3346 to IBC v3. I fixed with refference to these. I'm checking these.

Reply:
I'm afraid this won't create state breaking changes. Is it?

Reply:
https://github.com/cosmos/ibc-go/commit/74988e7b08427623bbf8b934aafcc6f341b96d1b
https://github.com/cosmos/ibc-go/blob/v4.1.2/CHANGELOG.md

https://github.com/cosmos/ibc-go/compare/v4.1.1...v4.1.2#diff-b225d8b78960f1ed19ccabcbdcda5ff708075790adac8463fbb0a83a8a63b78a
I have no idea because this fix does not have description. But this is included in update for huckleberry

Reply:
https://github.com/cosmos/ibc-go/releases/tag/v4.1.2
https://github.com/cosmos/ibc-go/compare/v4.1.2...v4.1.3
v4.1.2 seems to be not recommended because its version contain the state machine breaking change. But, I think moving event doesn't have problem because the update to v4.1.3 has nothing to do with it.

Reply:
We already have rervert experience as similar issue, Finschia/finschia-sdk#984. Therefore, I check it. But as you said, we didn't emit this events before. So, I also think it is no problem. But If we'll have some non-deterministic issue, I think we need to check this issue again. 

Thank you.